### PR TITLE
[babycare_partner_street_number][ADD] new extension module for street number addition

### PR DESCRIPTION
**Pull request by wytzejan**
- Introduces new field for street number addition.
- Disabled _write_street function because street fields may not be overwritten due to courier export.
- Rewrites of _get_street and _address_fields functions to include the street number.
- Include _display_address function to disable the same function in Webkul code.
- Rearranged fields in the partner form and added some translations.
- Added size=5 to existing field street_number because courier export only allows a 5-digit house number.

This extension modules depends on the forked OCA-module partner_street_number.

### CODE REVIEW

File: babycare_partner_street_number/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+{
+    "name": "Babycare Extension: Street name and number",
+    "summary": "Introduces new field for street number addition.",
+    "version": "8.0.0.1.0",
+    "author": "Wytze Jan Riedstra,Therp BV,Odoo Community Association (OCA)",
+    "category": 'Tools',
+    "depends": [
+        'partner_street_number',
```
Comment (StefanRijnhart) [inline babycare_partner_street_number/__openerp__.py:8]: You need to depend on wk_base_partner_patch to ensure that this module's _display_address works on that module's override of the same method.

File: babycare_partner_street_number/models/res_partner.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    @api.one
+    @api.depends('street_name', 'street_number', 'street_number_addition')
+    def _get_street(self):
+        self.street = ' '.join(
+            filter(None, [self.street_name, self.street_number,
+                          self.street_number_addition]))
+
+    def _write_street(self):
+        return False
```
Comment (StefanRijnhart) [inline babycare_partner_street_number/models/res_partner.py]: Any write on the 'street' field will now be dropped silently. Are you sure that this is safe? Possible data loss. You might want to raise an exception instead (although that might lead to blocking situations in your workflow). Having the original functionality from partner_street_number as a fallback still seems like a pretty good idea to me.

For existing addresses, you are too late removing the functionality here. Fortunately, the sanity limit applied on detected house numbers with a maximum length of 5 in the original method fits neatly to your length limit on that field.

File: babycare_partner_street_number/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        """
+        Inject a context key to prevent the 'street' name to be
+        deleted from the result of _address_fields when called from
+        the super.
+        """
+        return super(Partner, self.with_context(display_address=True)).\
+            _display_address(address, without_company=without_company)
+
+    @api.model
+    def _address_fields(self):
```
Comment (StefanRijnhart) [inline babycare_partner_street_number/models/res_partner.py:26]: ~~Do you need to repeat this method?~~ got it

File: babycare_partner_street_number/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        Could have checked for the occurrence of the synchronisation
+        method instead, leaving the field in by default but that could
+        lead to silent data corruption should the synchronisation API
+        ever change.
+        """
+        res = super(Partner, self)._address_fields()
+        if 'street' in res and not (
+                self._context.get('display_address')):
+            res.remove('street')
+        return res + ['street_name', 'street_number', 'street_number_addition']
```
Comment (StefanRijnhart) [inline babycare_partner_street_number/models/res_partner.py]: This module only adds 'street_number_addition'. You can leave out the fields from partner_street_number, as they are added in the super method.

**Comment by StefanRijnhart** [comment]
Looking good, overall!

File: babycare_partner_street_number/__openerp__.py (hunk 1 above)
Comment (wytzejan) [reply babycare_partner_street_number/__openerp__.py:8]: Fixed.

File: babycare_partner_street_number/models/res_partner.py (hunk 4 above)
Comment (wytzejan) [reply babycare_partner_street_number/models/res_partner.py]: Fixed.

File: babycare_partner_street_number/models/res_partner.py (hunk 3 above)
Comment (wytzejan) [reply babycare_partner_street_number/models/res_partner.py:26]: Great.

File: babycare_partner_street_number/models/res_partner.py (hunk 2 above)
Comment (wytzejan) [reply babycare_partner_street_number/models/res_partner.py]: I tested the original functionality from partner_street_number and this functionality is indeed great for the first import for example. But unfortunately it messes with the street, number and addition also afterwards. Let me give you an example.

- I created a Test Customer with the address Street (in street_name field) 32 (in street_number field) Bis (in street_number_addition field). babycare_partner_street_number_1

- After I hit save the _write_street method is 'styling' the address like this: 
babycare_partner_street_number_2

- As you can see in the edit-mode the 32 and Bis is added to the street in street_name but the 'Bis' is also still available in his own street_number_addition field. The street_number field is empty. 
babycare_partner_street_number_3

- This is the behavior after every change in one of the fields of street_name, street_number or street_number_addition. So for example if add a new number (36) in the street_number field but keep the existing number (32) in the street_name field the outcome is as follows:
babycare_partner_street_number_4

As you can see the _write_street method in the original module partner_street_number 'messes' with the outcome of the street fields. Therefore, I'd like to disable the _write_street method after the first install of the original module partner_street_number. I think my 'return False' proposal is a solution for this? But if you have a better solution I'd like to hear that!

Comment (wytzejan) [reply babycare_partner_street_number/models/res_partner.py]: Another important note: for the PostNL courier export it is necessary that all fields (street, street_number and street_number_addition (if applicable)) are filled. The rewrite is not usable in combination with the PostNL courier export.

Comment (StefanRijnhart) [reply babycare_partner_street_number/models/res_partner.py]: I see. It is a bug. In a form that properly displays the street name and street number fields, this method should not be called at all. It is purely meant as backwards compatibility for other addons that write to the street field directly. Attempted fix here: https://github.com/OCA/partner-contact/pull/478/files

Comment (StefanRijnhart) [reply babycare_partner_street_number/models/res_partner.py]: It's now https://github.com/OCA/partner-contact/pull/479

**Comment by StefanRijnhart** [comment]
Have you checked the fix for partner_street_number? You can remove your override of _write_street. I will make sure that the fix lands on the Odoo instances.

**Comment by wytzejan** [comment]
@StefanRijnhart I have checked the fix in partner_street_number and it works very good. Thanks!
I removed the _write_street override in my latest commit.

**Review by StefanRijnhart — APPROVED** [review]

**Comment by StefanRijnhart** [comment]
Thanks for the follow-up!

**Pull request merged**